app/assets/javascripts/discourse/app/lib/classic-component-events.js

```diff
@@ -0,0 +1,83 @@
+// eslint-disable-next-line ember/no-classic-components
+import Component from "@ember/component";
+import { EVENTS } from "../initializers/ember-events";
+
+/**
+ * @type {WeakMap<object, { event: string; method: string }[]>}
+ */
+const COMPONENT_SETUP = new WeakMap();
+
+const INTERNAL = Symbol("INTERNAL");
+
+/**
+ * Rewires classic component event handling to use `addEventListener` directly
+ * on inserted elements, instead of relying upon classic event delegation.
+ *
+ * This maximizes compatibility with glimmer components and event listeners
+ * added via the `on` modifier. In particular, using `addEventListener`
+ * consistently everywhere ensures that event propagation works as expected
+ * between parent and child elements.
+ *
+ * @param {Application} app
+ */
+export function rewireComponentEventHandlers(app) {
+  const allEvents = { ...EVENTS };
+
+  if (app.customEvents) {
+    for (const [event, methodName] of Object.entries(app.customEvents)) {
+      allEvents[event] = methodName;
+    }
+  }
+
+  Component.reopen({
+    /**
+     * @param {string | typeof INTERNAL} name
+     * @param {unknown[]} args
+     */
+    trigger(name, ...args) {
+      if (name === INTERNAL) {
+        return this._super.call(this, ...args);
+      } else if (name.toLowerCase() in allEvents) {
+        return;
+      } else {
+        return this._super.call(this, name, ...args);
+      }
+    },
+
+    initEventListeners() {
+      const proto = Object.getPrototypeOf(this);
+      let events = COMPONENT_SETUP.get(proto);
+
+      if (!events) {
+        events = [];
+        COMPONENT_SETUP.set(proto, events);
+
+        for (const [event, name] of Object.entries(allEvents)) {
+          if (this.has(name)) {
+            events.push({ event, name });
+          }
+        }
+      }
```

Comment:
Looks like we're building a cache based on the **prototype** of the component, but the list of events we're making is using `this.has()`, which can vary per component **instance** (e.g. if one instance of a component is passed an event handler as an argument, but another is not).

For example, if I make a simple component which extends from `@ember/component`, and then call it like this:

```handlebars
<SimpleComponent @click={{this.noOp}}>Simple1</SimpleComponent>
<SimpleComponent>Simple2</SimpleComponent>
```
Both of them are given `click` event handlers.

If I do it in the reverse order

```handlebars
<SimpleComponent>Simple1</SimpleComponent>
<SimpleComponent @click={{this.noOp}}>Simple2</SimpleComponent>
```
Neither of them get given event handlers

Reply:
Thanks for catching this @davidtaylorhq! I wasn't aware that this pattern was in use in the codebase, let alone how prevalent it is. I can easily see how this could lead to flakiness in the app and test suite.

I've added a [commit](https://github.com/discourse/discourse/pull/17767/commits/cac3a39257e741633745a94b65a5c640f4288753) to separately track instance vs. prototype event handlers. This separation adds a bit more overhead to each component instantiation, which is probably worth evaluating and optimizing further.

Reply:
Looks great to me, and CI is now green for plugins 🎉